crates/bitwarden-ssh/src/import.rs

```diff
@@ -0,0 +1,279 @@
+use ed25519;
+use pkcs8::{
+    der::Decode, EncryptedPrivateKeyInfo, ObjectIdentifier, PrivateKeyInfo, SecretDocument,
+};
+use serde::{Deserialize, Serialize};
+use ssh_key::{
+    private::{Ed25519Keypair, RsaKeypair},
+    HashAlg, LineEnding,
+};
+#[cfg(feature = "wasm")]
+use tsify_next::Tsify;
+
+use crate::{error::SshKeyImportError, SshKey};
+
+const PKCS1_HEADER: &str = "-----BEGIN RSA PRIVATE KEY-----";
+const PKCS8_UNENCRYPTED_HEADER: &str = "-----BEGIN PRIVATE KEY-----";
+const PKCS8_ENCRYPTED_HEADER: &str = "-----BEGIN ENCRYPTED PRIVATE KEY-----";
+const OPENSSH_HEADER: &str = "-----BEGIN OPENSSH PRIVATE KEY-----";
+
+pub const RSA_PKCS8_ALGORITHM_OID: ObjectIdentifier =
+    ObjectIdentifier::new_unwrap("1.2.840.113549.1.1.1");
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+enum KeyType {
+    Ed25519,
+    Rsa,
+    Unknown,
+}
+
+pub fn import_key(encoded_key: String, password: String) -> Result<SshKey, SshKeyImportError> {
+    match encoded_key.lines().next() {
+        Some(PKCS1_HEADER) => Err(SshKeyImportError::UnsupportedKeyType),
+        Some(PKCS8_UNENCRYPTED_HEADER) => {
+            import_pkcs8_key(encoded_key, None).map_err(|_| SshKeyImportError::ParsingError)
+        }
+        Some(PKCS8_ENCRYPTED_HEADER) => import_pkcs8_key(encoded_key, Some(password)),
+        Some(OPENSSH_HEADER) => import_openssh_key(encoded_key, password),
+        _ => Err(SshKeyImportError::ParsingError),
+    }
+}
+
+fn import_pkcs8_key(
+    encoded_key: String,
+    password: Option<String>,
+) -> Result<SshKey, SshKeyImportError> {
+    let der = match SecretDocument::from_pem(&encoded_key) {
+        Ok((_, doc)) => doc,
+        Err(_) => return Err(SshKeyImportError::ParsingError),
+    };
+
+    let decrypted_der = match password.clone() {
+        Some(password) => {
+            let encrypted_private_key_info = EncryptedPrivateKeyInfo::from_der(der.as_bytes())
+                .map_err(|_| SshKeyImportError::ParsingError)?;
+            encrypted_private_key_info
+                .decrypt(password.as_bytes())
+                .map_err(|_| SshKeyImportError::WrongPassword)?
```

Comment:
**thought:** `SecretDoc` defines a helper, `from_pkcs8_encrypted_pem` which could be used. `EncryptedPrivateKeyInfo` is somewhat lower level in my opinion?

You would need to map the error differently though since you get the generic pkcs8 error but it should expose a `DecryptFailed` case.

This combined with `SecretDocument::from_pem` suggestion above would let you lift in both in the match case.

Reply:
```rust
    let decrypted_der = if let Some(password) = password {
        SecretDocument::from_pkcs8_encrypted_pem(&encoded_key, password.as_bytes()).map_err(
            |err| match err {
                pkcs8::Error::EncryptedPrivateKey(pkcs5::Error::DecryptFailed) => {
                    SshKeyImportError::WrongPassword
                }
                _ => SshKeyImportError::ParsingError,
            },
        )?
    } else {
        SecretDocument::from_pkcs8_pem(&encoded_key).map_err(|_| SshKeyImportError::ParsingError)?
    };
```

Like this?

Reply:
Yes, that seems nice. We could probably even rename it to `der` now that we don't have a regular `der`.

Reply:
`SecretDocument::from_pkcs8_encrypted_pem` and `SecretDocument::from_pkcs8_pem` have the same result type. That allows you to move the error handling out of the nested code block.

Something like:
```rust
let der = if let Some(password) = password {
    SecretDocument::from_pkcs8_encrypted_pem(&encoded_key, password.as_bytes())
} else {
    SecretDocument::from_pkcs8_pem(&encoded_key)
};

let der = match.map_err(|err| match err {
    pkcs8::Error::EncryptedPrivateKey(pkcs5::Error::DecryptFailed) =>
        SshKeyImportError::WrongPassword,
    _ => SshKeyImportError::ParsingError
})?;
```

It works as a one-liner if you want it to, but since rust allows the variables to be rebound this seemed more clear.

Reply:
Agree, that does look cleaner.